### ISSUE TITLE
Refactor the compiler frontend

### ISSUE THREAD

**Issue by Michael-F-Bryan** [hotg-ai/rune#262]
There are two vague issues with the compiler frontend as it currently stands:

1. Accumulation of technical debt due to code churn/prototyping - makes it hard to follow the flow of logic and transformation of data
2. Ad-hoc state management - we've kinda got trees but not really, and the `Rune` type tends to just be a dumping ground of loosely/un related properties
3. Inability for 3rd parties to extend the `rune build` process - required for proprietary work

To deal with point 2, we can use an ECS library like [`legion`](https://docs.rs/legion) which gives us abstractions for working with entities ("things"), components (attributes we may want to attach to those things), and systems (chunks of behaviour for manipulating components or creating new entities). All state would be attached to entities and held in a world. 

Point 1 naturally falls out from this approach. Each compiler pass is just another system which attaches components to entities or takes one set of entities (e.g. syntax tree nodes) and derives another set of entities (e.g. pipeline nodes).

Point 3 is dealt with by providing a trait which has methods which get called at various steps along the compilation process. This is quite similar to the [Rustc Driver](https://rustc-dev-guide.rust-lang.org/rustc-driver.html) and how you can provide [callbacks](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/trait.Callbacks.html) that are fired at "interesting" points in the build process.

I could see the existing `hotg-rune-syntax` and `hotg-rune-codegen` crates being combined into a single `hotg-rune-compiler` crate which exposes something like this as its main interface:

```rust
use legion::World;

fn build(runefile: &str, context: &BuildContext, hooks: &mut dyn Hooks) -> World {
  let mut world = World::new();

  // Parse the YAML format and turn it into a syntax tree
  crate::passes::parse(&mut world, runefile);

  // Let the caller inspect our syntax tree and choose whether to continue
  hooks.after_parse(&world);

  // analyse the Runefile and resolve the pipeline
  crate::passes::register_stages(&mut world, context);
  crate::passes::register_register_output_slots(&mut world);
  crate::passes::construct_pipeline(&mut world);
  crate::passes::load_models_into_memory(&mut world);
  crate::passes::type_check(&mut world);

  // Let the caller inspect the resolved pipeline. This is where they might inspect
  // model node arguments and encrypt the loaded model
  hooks.after_analysis(&world);

  // Codegen
  crate::passes::generate_cargo_toml(&mut world, context);
  crate::passes::generate_lib_rs(&mut world);
  crate::passes::compile(&mut world, context);

  hooks.after_compile(&world);

  world
}

trait Hooks {
  fn after_parse(&mut self, _world: &World) {}
  fn after_analysis(&mut self, _world: &World) {}
  fn after_compile(&mut self, _world: &World) {}
}
```

CC: @jacekpie

**Comment by Michael-F-Bryan** [comment]
@kthakore, this is the refactor I was thinking of. 

It's a fairly sweeping change for our compiler frontend, but now we kinda know what form Rune will take I think it's time to go from the current ad-hoc procedural code to something more structured and extensible. We *could* try to hack something in that would let proprietary code update a Rune during the build process (e.g. to add crypto routines) without needing to patch the `syntax` crate directly, but that'd be quite messy in the long run.

**Comment by f0rodo** [comment]
Yeah this makes sense.
